lib/matplotlib/widgets.py

```diff
@@ -3885,37 +3871,36 @@ def _on_key_release(self, event):
                 and
                 (event.key == self._state_modifier_keys.get('move_vertex')
                  or event.key == self._state_modifier_keys.get('move_all'))):
-            self._xs.append(event.xdata)
-            self._ys.append(event.ydata)
+            self._xys.append((event.xdata, event.ydata))
             self._draw_polygon()
         # Reset the polygon if the released key is the 'clear' key.
         elif event.key == self._state_modifier_keys.get('clear'):
             event = self._clean_event(event)
-            self._xs, self._ys = [event.xdata], [event.ydata]
+            self._xys = [(event.xdata, event.ydata)]
             self._selection_completed = False
             self._remove_box()
             self.set_visible(True)
 
     def _draw_polygon(self):
         """Redraw the polygon based on the new vertex positions."""
-        self._selection_artist.set_data(self._xs, self._ys)
+        xs, ys = zip(*self._xys) if self._xys else ([], [])
+        self._selection_artist.set_data(xs, ys)
         self._update_box()
         # Only show one tool handle at the start and end vertex of the polygon
         # if the polygon is completed or the user is locked on to the start
         # vertex.
         if (self._selection_completed
-                or (len(self._xs) > 3
-                    and self._xs[-1] == self._xs[0]
-                    and self._ys[-1] == self._ys[0])):
-            self._polygon_handles.set_data(self._xs[:-1], self._ys[:-1])
+                or (len(self._xys) > 3
+                    and self._xys[-1] == self._xys[0])):
+            self._polygon_handles.set_data(xs[:-1], ys[:-1])
         else:
-            self._polygon_handles.set_data(self._xs, self._ys)
+            self._polygon_handles.set_data(xs, ys)
         self.update()
 
     @property
     def verts(self):
         """The polygon vertices, as a list of ``(x, y)`` pairs."""
-        return list(zip(self._xs[:-1], self._ys[:-1]))
+        return self._xys[:-1]
```

Comment:
This now returns a handle to the internal data structure, which may have unintended side effects and could even break existing code. Should we return a copy?

Reply:
Actually `self._xys[:-1]` returns a copy of the `self._xys` list (it's not a ndarray).

Reply:
Oh, yes of course.